### PR TITLE
feat: Ability to relate private documents to actors

### LINKED ISSUE #2762: Ability to share a resource of  DAC (document access control)
Ability to share document access through acp

### PR DESCRIPTION
**Pull request by shahzadlone**
## Relevant issue(s)
Resolves #2762

## Description
This PR introduces the ability to make use of the `relation`s defined within a policy to create relationships between an actor and a document within a collection. For users sake, I have made the clients (http, and cli) not consume the `policyID` and `resource` name but instead a `docID` and `collection name`, since the collection will have the policy and resource information available we can fetch that and make lives easier for the users.

This PR also makes use of the `manages` feature we have had in our policy. The manages essentially defines who can make the relationship manipulation requests.

There are a lot of tests in this PR due to a lot of edge cases I wanted to have tested specific to `manger`, and ensuring `write` and `read` permissions don't leak (i.e. are accidently granted).

## CLI Demo
The following lets the target actor be able to now read the private document: 

```bash
defradb client acp relationship add \
--collection Users \
--docID bae-ff3ceb1c-b5c0-5e86-a024-dd1b16a4261c \
--relation reader \
--actor did:key:z7r8os2G88XXBNBTLj3kFR5rzUJ4VAesbX7PgsA68ak9B5RYcXF5EZEmjRzzinZndPSSwujXb4XKHG6vmKEFG6ZfsfcQn \
--identity e3b722906ee4e56368f581cd8b18ab0f48af1ea53e635e3f7b8acd076676f6ac
```

Result:
```json
{
  "ExistedAlready": false // <-------------- Indicates a new relationship was formed
}
```

### Future (out-of-scope of this PR):
- Most of write tests will split into `delete` and `update` in #2905 
- Ability to revoke or delete relation coming in #2906 
- Decide on the `can't write if no read permission` in #2992 
- Move acp logic to a shared repo: https://github.com/sourcenetwork/defradb/issues/2980

## How has this been tested?
- Integration tests

Specify the platform(s) on which this was tested:
- Manjaro WSL2

### CODE REVIEW

File: tests/integration/acp.go
Comment (shahzadlone) [inline tests/integration/acp.go:184]: discussion: I understand if people prefer having `DocIDIndex` and `DocID` to test these edge cases rather than `-1` to test empty/invalid DocID input. I don't have any strong preferences here, happy to follow consensus.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_only_write_test.go
Comment (shahzadlone) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_only_write_test.go:51]: discussion: Since in these tests writer won't get `read` `permission`, even if a relation is made to make an actor a writer, they won't be able to write because the policy doesn't model ability to read. This behavior we can try manipulate on defradb level to change (however not in this PR), but opening this here to start a discussion.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_gql_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_gql_test.go:552]: question: I don't see how this triggers the error. The title of the test says "CanNotGivesAccessNotDefinedInPolicy", but "writer" IS defined in the policy.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:209]: type: ManagerGivesWriteAccess

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: question: why not to make manager automatically have this permissions? It seems to be pointless to do it separately.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:845]: suggestion: I think it should be "TestACP_ManagerGivesAccessNotDefinedInPolicyToAnotherActor_Error"

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:1162]: typo: manger?

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go
Comment (islamaliev) [inline tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go:306]: question: why does this need to be tested?

**Review by islamaliev — COMMENTED** [review]
Good work so far. I'm just not sure why it's still in Draft state.
There are a lot of tests here, but I'm not sure if all them are necessary.
A lot of repetitive code in tests, which makes it hard to read every detail.

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:184]: `immutable.Option` is for optional fields. This is not an optional field, it's more to test invalid input of that field. Is why I was wondering if others prefer non-index input as well (i.e. `DocIDIndex` for normal cases and `DocID` for actual invalid testing).

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_delete_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_delete_test.go:281]: - [ ] Revisit and remove some earlier steps that are tested already in other tests

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_gql_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_gql_test.go:552]: That is because a manager can not give access to another actor with a relation that it does not manage.

Good shout on the test name, will change to something better.

- [ ] Better test name

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: This is a sourcehub level decision not a defradb one. I am guessing it was for more fine grain metrics or flexibility in future.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go:306]: There is an operation on sourcehub called `transfer ownership` which does exactly that. So here I wanted to make sure no bugs slip through in the future. i.e. if the `sharing` ever `transfers` the document ownership it should definitely be caught.

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:123]: - [x] done

File: tests/integration/acp.go
Comment (shahzadlone) [reply tests/integration/acp.go:133]: - [x] Done for this and other fields

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:209]: - [x] Fix typo

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:1162]: - [x] Done

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (islamaliev) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: the questions is then why are we testing sourcehub functionality so vigorously on defra side? After all, this is just one of potential remote ACP implementations. What if other ACPs don't have such behavior? Or what if it changes on the sourcehub side?

Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: > the questions is then why are we testing sourcehub functionality so vigorously on defra side?
I don't consider this as `vigorous` at all. It's just that the manager functionality existed before since acp was implemented but it doesn't become functional or even testable until we have relationship feature. So now that we have the relationship sharing implemented this is important to cover in my opinion as it's something users can now use.

> After all, this is just one of potential remote ACP implementations.
Not sure I understand, but every test runs against both `local` and `sourcehub` implementation. 

> What if other ACPs don't have such behavior?
What is this other ACP you are referring to? Not sure I follow.

> Or what if it changes on the sourcehub side?
Then that is a change that DefraDB must adjust with like with all other third party dependency changes (however I doubt this behavior will change).

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go:306]: I can add a comment here sure, but I don't see how this makes hard for other non-sourcehub implementations. They should all have the same behavior (in this case making sure that after new actor can access a document the previous owner doesn't loose access, i.e. is still the owner). I understand is an edge case, but this will be true for all acp implementations I can think of

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:845]: - [x] Changed to a better name: `TestACP_ManagerAddsRelationshipWithRelationItDoesNotManageAccordingToPolicy_Error`

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_delete_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_delete_test.go:133]: While I like the suggestion, this seems like a personal preference unless we reach consensus as a team to follow this rule. So leaving as is.

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go
Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_reader_test.go:308]: - [x] Added comment

File: tests/integration/acp/p2p/replicator_with_doc_actor_relationship_test.go
Comment (AndrewSisley) [inline tests/integration/acp/p2p/replicator_with_doc_actor_relationship_test.go:144]: question: I've seen you do this quite a bit, why do you add whitespace between every property in an instance declaration?

Comment (shahzadlone) [reply tests/integration/acp/p2p/replicator_with_doc_actor_relationship_test.go:144]: No good reason, I am claustrophobic perhaps and like spaced lines lol. I often wonder why people don't add whitespaces.

File: acp/source_hub_client.go
Comment (AndrewSisley) [inline acp/source_hub_client.go:409]: todo: Duplicating the log between the if-else here really reduces the information density, and increases the amount of stuff maintainers have to read. Please simplify this to something similar to the below:
```
log.InfoContext(
			ctx,
			"Document and actor relationship set",
			corelog.Any("PolicyID", policyID),
			corelog.Any("ResourceName", resourceName),
			corelog.Any("DocID", docID),
			corelog.Any("Relation", relation),
			corelog.Any("RequestActor", requestActor.DID),
			corelog.Any("TargetActor", targetActor),
			corelog.Any("Existed", exists),
		)
		return exists, nil
```

Comment (shahzadlone) [reply acp/source_hub_client.go:409]: Good suggestion, will do.

- [ ] logs

File: tests/integration/acp/p2p/replicator_with_doc_actor_relationship_test.go
Comment (shahzadlone) [reply tests/integration/acp/p2p/replicator_with_doc_actor_relationship_test.go:144]: TBH it's just a habit while writing code, I don't care if some formatter formats these style choices to one format that is more consistent (idk if there is one, or a linter rule?)

File: tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go
Comment (AndrewSisley) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: >  why are we testing sourcehub functionality so vigorously on defra side?

I think it is worth having a proper team conversation about this - the ACP interface is kind of a black-box to Defra, and it should be up to the implementations to ensure they properly satisfy the interface.  Defra just needs to worry about calling correctly.

The maintenance cost for all these tests is not free.

Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: > > why are we testing sourcehub functionality so vigorously on defra side?
> 
> I think it is worth having a proper team conversation about this - the ACP interface is kind of a black-box to Defra, and it should be up to the implementations to ensure they properly satisfy the interface. Defra just needs to worry about calling it correctly.
> 
> The maintenance cost for all these tests is not free.

I agree, I do kind of look at this situation as along the same line as `Unit test` vs `Integration test` specially because the "sourcehub unit" at this current stage lacks the level of testing I would feel comfortable with specially when using it to deliver some key privacy/security feature of a database.

Comment (AndrewSisley) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: > the "sourcehub unit" at this current stage lacks the level of testing

Yes, this is a problem, but I very strongly believe that any Source member should be free to correct that instead of over-compensating in what could be seen as the wrong repo.

Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: > Source member should be free to correct that instead of feeling like they need to over-compensate in what could be seen as the wrong repo.

This is actually a very good idea, this way the testing can belong in more appropriate areas. For example the DPI stuff `write` and `read` specific permission testing can happy on defra and the general invalid and edge cases can belong on sourcehub.

Comment (islamaliev) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: I know it feels more reassuring to have many tests covered on Defra side, but as Andy pointed out, it comes at a cost. Defra should test only defra-related edge cases and some general happy-paths to make sure modules work well together.

Comment (islamaliev) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: > What is this other ACP you are referring to? Not sure I follow.

We are thinking for example about implementing UCAN-based remote ACP which relies purely on decentralized permission resolution between peers, i.e. without some entity (like sourcehub) in the middle.

Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: I am happy to follow consensus, can point to individual tests that are seemingly out of defra-scope and I will push back only if I do deem it "defra" specific, otherwise will try to move them to sourcehub :)

Comment (shahzadlone) [reply tests/integration/acp/relationship/add_doc_actor_test/add_doc_actor_with_manager_test.go:410]: Tracking here: https://github.com/sourcenetwork/defradb/issues/2980

**Review by islamaliev — APPROVED** [review]
Approving assuming tests will be moved as discussed

File: internal/db/db.go
Comment (AndrewSisley) [inline internal/db/db.go:213]: todo: This is already handled by `GetCollectionByName` which will error if no collection is found.  Please remove the new error and this block.

File: client/db.go
Comment (AndrewSisley) [inline client/db.go:106]: > // Moreover, since collection has the policy and resource information we don't need to
	// explicitly have the user provide them as arguments

todo: I'm not sure I understand this, can you please reword it?  Are some of these params optional?

File: acp/source_hub_client.go
Comment (AndrewSisley) [inline acp/source_hub_client.go:368]: suggestion: This comment is not very useful, please either expand it to cover *why* all of the are required (and move it to the func-level), or remove it - the code already states that they are required, without a why this comment is not useful.

Comment (shahzadlone) [reply acp/source_hub_client.go:368]: - [ ] remove redundant comment

File: internal/db/db.go
Comment (shahzadlone) [reply internal/db/db.go:213]: good catch

- [x] removed

File: client/db.go
Comment (shahzadlone) [reply client/db.go:106]: params are not optional was trying to explain that the magic of having policy and resource info on collection is nice because the user doesn't have to input that but just the collection name. going to remove it as it causes more confusion that clarification

- [x] extra fluff makes things more confusing to remove

**Review by AndrewSisley — APPROVED** [review]
LGTM, thanks Shahzad!

**Pull request merged**
